keras/layers/core.py

```diff
@@ -995,6 +995,86 @@ def get_config(self):
         return dict(list(base_config.items()) + list(config.items()))
 
 
+class TimeDistributed(MaskedLayer):
+    def __init__(self, layer, input_shape=None, input_dim=None, input_length=None, weights=None, **kwargs):
+        self.layer = layer
+
+        self.initial_weights = weights
+
+        self.input_dim = input_dim
+        self.input_length = input_length
+
+        if hasattr(self.layer, 'input_ndim'):
+            self.input_ndim = self.layer.input_ndim + 1
+
+        if input_shape:
+            self.set_input_shape((None, ) + input_shape)
+
+        if self.input_dim:
+            kwargs['input_shape'] = (self.input_length, self.input_dim)
+
+        super(TimeDistributed, self).__init__(**kwargs)
+
+    def set_previous(self, layer):
+        self.input_ndim = len(layer.output_shape)
+        super(TimeDistributed, self).set_previous(layer)
+
+    def build(self):
+        try:
+            self.input_ndim = len(self.previous.input_shape)
+        except AttributeError:
+            self.input_ndim = len(self.input_shape)
+
+        self.layer.set_input_shape((None, ) + self.input_shape[2:])
```

Comment:
The batch size dimension might not be `None`. It has to be specified in quite a few situations (e.g. usage of the present layer in a stateful network).